pkg/kotsadm/minio.go

```diff
@@ -97,12 +121,20 @@ func ensureMinioStatefulset(deployOptions types.DeployOptions, clientset *kubern
 	existingMinio.Spec.Template.Spec.Volumes = desiredMinio.Spec.Template.Spec.DeepCopy().Volumes
 	existingMinio.Spec.Template.Spec.Containers[0].Image = desiredMinio.Spec.Template.Spec.Containers[0].Image
 	existingMinio.Spec.Template.Spec.Containers[0].VolumeMounts = desiredMinio.Spec.Template.Spec.Containers[0].DeepCopy().VolumeMounts
+	existingMinio.Spec.Template.Spec.InitContainers = desiredMinio.Spec.Template.Spec.DeepCopy().InitContainers
 
 	_, err = clientset.AppsV1().StatefulSets(deployOptions.Namespace).Update(ctx, existingMinio, metav1.UpdateOptions{})
 	if err != nil {
 		return errors.Wrap(err, "failed to update minio statefulset")
 	}
 
+	// ensureMinioStatefulset can return before the init containers are actually added or removed from the pods
+	// if this occurs, the statefulset will show as a false 'Ready' status and kotsadm will start before the migration is complete
+	desiredInitContainers := len(existingMinio.Spec.Template.Spec.InitContainers)
+	if err := waitForMinioInitContainers(ctx, deployOptions.Namespace, clientset, deployOptions.Timeout, desiredInitContainers); err != nil {
```

Comment:
doesn't `WaitForStatefulSetReady` take care of this?

Reply:
not always. I observed this function returning before the minio statefulset was updated and ready ready.  it seems that there is a race condition where if the statefulset pods have not begun updating when the check happens, it will see it as ready (which it is, but we need it to actually update before the check happens).  i think it may be because we now update the initcontainers, so k8s might be taking a bit longer to process the change and issue the update

Reply:
the `WaitForStatefulSetReady` checks the `generation` and `observedGeneration` which are immediately updated even if the pods did not roll yet. so, the only race condition would happen if the actual statefulset definition is not updated yet, which seems a bit unlikely.

Reply:
The check does not seem to be working as expected, so I guess I'll have to dig in more to understand why.

Reply:
Minio StatefulSet is "Ready" and has the updated image, but the pods have not restarted:
<img width="1162" alt="Screen Shot 2023-02-21 at 11 32 04 AM" src="https://user-images.githubusercontent.com/17422963/220404598-d7ca22e5-7e5e-4568-8d20-6661431d5d26.png">

Full loom (this occurs at 10s): https://www.loom.com/share/cb7c89a59a2e45eb8704177943ddfb88

Reply:
resolved: https://github.com/replicatedhq/kots/pull/3624#discussion_r1113429594